include/onnxruntime/core/session/onnxruntime_cxx_inline.h

```diff
@@ -425,6 +425,11 @@ inline Env& Env::CreateAndRegisterAllocator(const OrtMemoryInfo* mem_info, const
   return *this;
 }
 
+inline void detail::LogImpl(OrtLoggingLevel severity, const char* message, const char* file_path, int line_number,
+                            const char* func_name) {
+  ThrowOnError(GetApi().Log(severity, message, file_path, line_number, func_name));
```

Comment:
>ThrowOnError [](http://example.com/codeflow?start=2&length=12)

Throwing on Logging makes it impossible to log within a catch clause or from a destructor.

Reply:
I am not sure we could deal with exceptions in a way of enumerating possible causes. C++ can throw anything from anywhere. Typically, when a function is not trivial, like the above API entry point, they only way to deal with it is to wrap it in `try/catch` and that is what BEGIN/END accomplish, so C++ exceptions do not propagate the C API boundaries.
My take on it, we can still report them via C API, but do not translate it to C++ exceptions, since we have no way of dealing with them in this C++ context. In fact, the C API cannot log either since it errors out. Let's make it noexcpt since destructors may use it.